fix clippy::unwrap_used in re_viewer with Option::expect

brody4hire opened the pull request:
### Related

Part of: #6330

### What

- remove allow directive & TODO comment from `crates/viewer/re_viewer/src/lib.rs`
- use `. unwrap_or_else` with panic in 1 case
- use focused `#[allow(clippy::unwrap_used)]` directives in limited, focused cases
- use `.expect` with descriptive messages in all other cases

(no changelog updates)

__General rationale:__

Most of the cases using `.unwrap` seem to need these to work. I think the best solution is to use `.expect` with a clear message in most of these cases.

__Sanity checks done:__

These succeed for me with no warnings:

- `cargo clippy -p re_viewer --all-features`
- `cargo build -p re_viewer --all-features`
- `cargo fmt` does not introduce inconsistent formatting of the updated code

Wumpf reviewed (approved):
thank you!!

brody4hire commented on crates/viewer/re_viewer/src/app_blueprint.rs:
```diff
@@ -219,10 +220,12 @@ impl AppBlueprint<'_> {
 
             let timepoint = store_ctx.blueprint_timepoint_for_writes();
 
+            // Can only fail if we have the wrong number of instances for the component, and we don't
```
~~need to clean this comment as well~~

---

I would like to remove this comment & `#[allow(clippy::unwrap_used)]` in favor of using `.expect()` here as well.

brody4hire commented on crates/viewer/re_viewer/src/app_blueprint.rs:
```diff
@@ -197,11 +197,11 @@ pub fn setup_welcome_screen_blueprint(welcome_screen_blueprint: &mut EntityDb) {
         welcome_screen_blueprint
             .add_chunk(&Arc::new(chunk))
-            .unwrap(); // Can only fail if we have the wrong number of instances for the component, and we don't
+            .expect("Failed to add new chunk for welcome screen - internal error");
```
nit: improve consistency here

```suggestion
            .expect("Failed to add new chunk for welcome screen (internal error)");
```

brody4hire reviewed (commented):
@Wumpf I have shamelessly thrown away your cleanup work in favor of using `.expect("... (internal error)")` calls (I am assuming no offense will be taken). I think this should both reduce the number of lines in most cases and clarify the code a little better.

I think this should be ready now. Please do feel free to apply any other cleanup you think may be needed & be sure to add yourself as a co-author.

emilk commented on crates/viewer/re_viewer/src/ui/welcome_screen/welcome_section.rs:
```diff
@@ -56,7 +56,9 @@ pub(super) fn welcome_section_ui(ui: &mut egui::Ui) {
                 egui::RichText::new("Go to documentation →")
-                    .color(egui::Color32::from_hex("#60A0FF").expect("this color is valid"))
+                    .color(
+                        egui::Color32::from_hex("#60A0FF").expect("color failure (internal error)"),
```
As noted in other PR: This should be using the `egui::hex_color!` macro instead: https://docs.rs/egui/latest/egui/macro.hex_color.html

emilk commented on crates/viewer/re_viewer/src/screenshotter.rs:
```diff
@@ -89,7 +89,7 @@ impl Screenshotter {
             let image =
                 image::RgbaImage::from_raw(w, h, bytemuck::pod_collect_to_vec(&image.pixels))
-                    .expect("Failed to create image");
+                    .expect("Failed to create image (internal error)");
```
I like the "internal error" message for our users, but I think we should add that to the panic handler of `re_crash_handler` instead, since any panic is an internal error.

brody4hire replied:
I would agree in most but not all cases, I think this points to a counter-example as I commented below: https://github.com/rerun-io/rerun/pull/8745/files#r1928954634

brody4hire replied:
How do you think we should proceed with this?

emilk replied:
A panic _should_ only happen because of a bug in the rerun codebase, which is what I would call an internal error. If there are panics that could happen because of user error, we should fix that separatedly.

brody4hire commented on crates/viewer/re_viewer/src/background_tasks.rs:
```diff
@@ -60,7 +60,10 @@ impl BackgroundTasks {
-                Ok(any) => Some(*any.downcast::<T>().unwrap()),
+                Ok(any) => Some(
+                    *any.downcast::<T>()
+                        .unwrap_or_else(|err| panic!("downcast failure: {err:?}")),
```
I think this is an example where a panic is triggered _not_ by an internal error but by an issue with the calling code.

brody4hire replied:
I am thinking it may be ideal to update the API to return a promise or result with a promise so that the calling code can poll it without the need for this downcast. We could add a separate API function to do this without breaking any API users. WDYT?

brody4hire commented on crates/viewer/re_viewer/src/app.rs:
```diff
@@ -1836,7 +1836,10 @@ impl eframe::App for App {
 
         // Temporarily take the `StoreHub` out of the Viewer so it doesn't interfere with mutability
-        let mut store_hub = self.store_hub.take().unwrap();
+        let mut store_hub = self
+            .store_hub
+            .take()
+            .expect("Failed to take store hub");
```
D'oh `cargo fmt` flattens this now that I shortened the expect message, which is not good consistency with the other updates I made and not what I really wanted to happen here - argh

brody4hire commented:
I have rebased (2x), removed "` (internal error)`" from the end of the expect messages, and updated an error message to keep `cargo fmt` from flatting an update. Please let me know if anything else is needed from me on this.

emilk reviewed (approved):
Thank you for this!
